software/contrib/conway.py

```diff
@@ -0,0 +1,354 @@
+        # the size of the field
+        self.num_changes = 0
+        self.changed_spaces = bytearray(NUM_PIXELS >> 3)
+        self.next_changed_spaces = bytearray(NUM_PIXELS >> 3)
+
+        # how many cells were born this tick?
+        self.num_born = 0
+
+        # how many cells died this tick?
+        self.num_died = 0
+
+        # how many cells are currently alive?
+        # this gets updated on every tick and on random spawns
+        self.num_alive = 0
+
+        # Set to True if we want to clear the field & respawn
+        self.reset_requested = False
+
+        # keep the last few changes in population in a list to check if it's oscillating predictably
+        self.population_deltas = []
+        self.MAX_DELTAS = 12
+
+        @b1.handler
+        def on_b1():
+            self.reset_requested = True
+
+        @b2.handler
+        def on_b2():
+            self.reset_requested = True
+
+        @din.handler
+        def on_din():
+            self.reset_requested = True
+
+    def get_neigbour_indices(self, index):
+        """Get the indices of the 8 bits adjacent to the given index
+
+        If we're on the top/left/bottom/right edge, wrap arround to the opposite row/column, treating the world
+        as a torus
+
+        Unfortunately we don't have enough RAM to save this in a lookup table, so we have to recalculate it
+        every time, which slows down the simulation
+        """
+        row = index // OLED_WIDTH
+        col = index % OLED_WIDTH
+
+        def rowcol2index(r, c):
+            return (r % OLED_HEIGHT) * OLED_WIDTH + (c % OLED_WIDTH)
+
+        neighbours = []
+
+        for i in range(-1, 2):
+            for j in range(-1, 2):
+                if i != 0 or j != 0:
+                    neighbours.append(rowcol2index(row+i, col+j))
+
+        return neighbours
+
+    def calculate_spawn_level(self):
+        """Calculate what percentage of the field should contain new cells
+
+        We want to avoid having the spawn rate _too_ high as that can result in high CPU loads and RAM usage.
+        - K1: [0, 0.5]
+        - K2: [-0.5, 0.5]
+        - AIN: [0, 1]
+
+        P = K1 + K2 * AIN => [-0.5, 1.0]
+        """
+        base_spawn_level = k1.percent() / 2
+
+        # get the level of AIN, attenuverted by K2
+        cv_mod = ain.percent()/2
+        cv_att = k2.percent() - 1
+        cv_mod = cv_mod * cv_att
+
+        spawn_level = clamp(base_spawn_level + cv_mod, 0, 1)
+        return spawn_level
+
+    def reset(self):
+        """Clear the whole field and spawn random data in it
+        """
+        for i in range(len(self.field)):
+            self.field[i] = 0x00
+            self.next_field[i] = 0x00
+
+        self.num_alive = 0
+        self.population_deltas = []
+
+        # fill the field with random cells
+        fill_level = self.calculate_spawn_level()
+        for i in range(NUM_PIXELS):
+            x = rnd()
+            is_alive = get_bit(self.field, i)
```

Comment:
[question] I'm trying to understand the approach here and am having trouble. It looks like you first clear the entire filed. Then you iterate over the pixels in order and decide if it should be alive based on a random number and the fill level. In this case wouldn't `is_alive` here always be false?

Reply:
Yes, you're correct.  The field should in fact not be cleared before this, just `next_field`.  The idea here was to try to speed things up only by adding/removing cells where needed.